# Grupo: 

*   Maria Luiza Vieira Belo 
*   Lucca Spinelli
*   Marcello Augusto
*   Eliziane Mota
*   Jardel Simplício 

- Turma de ADS Regular 3º Período


# Atividade Prática: Estatística Descritiva e Exploratória

Nesta atividade prática, vamos explorar os conceitos fundamentais da Estatística Descritiva e Exploratória (EDA). Utilizaremos o famoso dataset 'Tips' da biblioteca `seaborn`, que contém informações sobre gorjetas dadas em um restaurante. O objetivo é que vocês apliquem os conhecimentos teóricos na prática, manipulando dados, calculando métricas e criando visualizações para extrair insights valiosos.

### Objetivos de Aprendizagem:

*   Carregar e inspecionar um dataset.
*   Calcular e interpretar medidas de tendência central e dispersão.
*   Identificar e analisar outliers.
*   Avaliar a forma da distribuição de dados (assimetria e curtose).
*   Analisar a correlação entre variáveis numéricas.
*   Criar visualizações essenciais como histogramas e boxplots.

In [ ]:
# Importando as bibliotecas necessárias
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from scipy.stats import skew, kurtosis

# Carregando o dataset 'Tips'
df = sns.load_dataset('tips')

print("Dataset 'Tips' carregado com sucesso!")

Dataset 'Tips' carregado com sucesso!


## 1. Exploração Inicial do Dataset

Antes de mergulharmos em análises mais complexas, é crucial entender a estrutura básica dos nossos dados. A exploração inicial nos ajuda a ter uma primeira impressão das variáveis, seus tipos, valores ausentes e o tamanho geral do dataset. As funções `head()`, `info()` e `shape` são excelentes para isso.

### Exercício 1.1: Visão Geral dos Dados

Utilize os métodos `head()`, `info()` e o atributo `shape` para obter uma visão geral do dataset `df`. Observe os tipos de dados de cada coluna, a presença de valores nulos e o número de linhas e colunas.

In [ ]:
df.head()

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4


In [ ]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   total_bill  244 non-null    float64 
 1   tip         244 non-null    float64 
 2   sex         244 non-null    category
 3   smoker      244 non-null    category
 4   day         244 non-null    category
 5   time        244 non-null    category
 6   size        244 non-null    int64   
dtypes: category(4), float64(2), int64(1)
memory usage: 7.4 KB


In [ ]:
df.shape

(244, 7)

### Análise do Aluno:

O dataset refere-se a dados de um restaurante, contendo informações sobre contas `total_bill` e gorjetas `tip`

O conjunto de dados é relativamente pequeno, com 244 registros (linhas) e 7 variáveis (colunas). Não há valores nulos, pois todas as colunas possuem 244 valores não nulos (non-null), indicando que o dataset está completo. 

Os tipos de dados são:

- Númericos:
  - `total_bill` → float
  - `tip` → float
  - `size` → int
- Categóricos
  - `sex`, `smoker`, `day`, `time`

Essa divisão permite realizar:

- análises estatísticas sobre os dados numéricos
- segmentações e comparações com base nas variáveis categóricas

Além disso, o dataset possui um baixo consumo de memória (7.4 KB). O uso do tipo `category` nas variáveis categóricas contribui para essa eficiência, indicando uma boa preparação dos dados.


## 2. Medidas de Tendência Central e Dispersão

As medidas de tendência central (média, mediana) nos indicam o 'centro' dos nossos dados, enquanto as medidas de dispersão (desvio padrão, variância) nos dizem o quão espalhados esses dados estão em torno desse centro. Elas são fundamentais para entender a distribuição de uma variável.

### Exercício 2.1: Cálculo de Medidas Centrais e de Dispersão

Para as colunas numéricas ('total_bill', 'tip', 'size'), calcule a média, mediana, desvio padrão e variância. Utilize o método `describe()` do pandas para uma visão rápida e depois calcule cada uma individualmente.

In [ ]:
df.describe()

,total_bill,tip,size
count,244.000000,244.000000,244.000000
mean,19.785943,2.998279,2.569672
std,8.902412,1.383638,0.951100
min,3.070000,1.000000,1.000000
25%,13.347500,2.000000,2.000000
50%,17.795000,2.900000,2.000000
75%,24.127500,3.562500,3.000000
max,50.810000,10.000000,6.000000


#### Coluna: `total_bill`

In [ ]:
col_name = 'total_bill'
print(f"Coluna {col_name}:")
print(f"Mean: {df[col_name].mean()}")
print(f"Mediana: {df[col_name].median()}")
print(f"Desvio Padrão: {df[col_name].std()}")
print(f"Variância: {df[col_name].var()}")

Coluna total_bill:
Mean: 19.78594262295082
Mediana: 17.795
Desvio Padrão: 8.902411954856856
Variância: 79.25293861397827


#### Coluna: `total_bill`
- média: 19.78
- mediana: 17.79

média > mediana

- desvio padrão: 8.90
- variância: 79.25

Alta variabilidade e Valores bem espalhados

#### Coluna: `tip`

In [ ]:
col_name = 'tip'
print(f"Coluna {col_name}:")
print(f"Mean: {df[col_name].mean()}")
print(f"Mediana: {df[col_name].median()}")
print(f"Desvio Padrão: {df[col_name].std()}")
print(f"Variância: {df[col_name].var()}")

Coluna tip:
Mean: 2.99827868852459
Mediana: 2.9
Desvio Padrão: 1.3836381890011822
Var: 1.914454638062471


#### Coluna: `tip`
- média: 2.99
- mediana: 2.9

média > mediana (diferença pequena)

- desvio padrão: 1.38
- variância: 1.91

Variabilidade moderada

#### Coluna: `size`

In [ ]:
col_name = 'size'
print(f"Coluna {col_name}:")
print(f"Mean: {df[col_name].mean()}")
print(f"Mediana: {df[col_name].median()}")
print(f"Desvio Padrão: {df[col_name].std()}")
print(f"Variância: {df[col_name].var()}")

Mean: 2.569672131147541
Mediana: 2.0
Desvio Padrão: 0.9510998047322344
Var: 0.9045908385616945


#### Coluna: `size`
- média: 2.57
- mediana: 2.0

média > mediana

- desvio padrão: 0.95
- variância: 0.90

Baixa variabilidade

### Análise do Aluno:
A comparação entre média e mediana mostra que todas as variáveis apresentam assimetria à direita, já que a média é maior que a mediana em todos os casos. Isso indica a presença de valores mais altos que puxam a média para cima, especialmente em `total_bill`.

Em relação à variabilidade, `total_bill` apresenta alta dispersão, indicando grande variação nos valores das contas. A variável `tip` possui variabilidade moderada, enquanto `size` apresenta baixa dispersão, mostrando que o tamanho dos grupos é mais consistente.

## 3. Análise de Quartis e Outliers

Os quartis dividem um dataset ordenado em quatro partes iguais (Q1: 25%, Q2: 50% - mediana, Q3: 75%). Eles são cruciais para entender a dispersão e identificar possíveis outliers. Um outlier é um ponto de dado que se desvia significativamente de outras observações. Podemos identificá-los usando o Intervalo Interquartil (IQR), que é a diferença entre Q3 e Q1, e estabelecendo limites inferior e superior.

### Exercício 3.1: Cálculo Manual de Quartis e Identificação de Outliers

Para a coluna 'total_bill', calcule manualmente (sem usar `describe()` diretamente para os quartis, mas pode usar `quantile()`) os valores de Q1, Q2 (mediana), Q3 e o Intervalo Interquartil (IQR). Em seguida, calcule o Limite Inferior e Superior para identificação de outliers (Limite Inferior = Q1 - 1.5 * IQR; Limite Superior = Q3 + 1.5 * IQR). Por fim, mostre quantos outliers existem na coluna 'total_bill'.

In [ ]:
df = sns.load_dataset('tips')
print("Dataset 'Tips' carregado com sucesso!")

q1 = df['total_bill'].quantile(0.25)
q2 = df['total_bill'].median()
q3 = df['total_bill'].quantile(0.75)
iqr = q3 - q1

limite_inferior = q1 - 1.5 * iqr
limite_superior = q3 + 1.5 * iqr

outliers = df[(df['total_bill'] < limite_inferior) | (df['total_bill'] > limite_superior)]

print(f"\n--- Resultados do Exercício 3.1 ---")
print(f"Q1: {q1:.2f} | Q3: {q3:.2f} | IQR: {iqr:.2f}")
print(f"Limite Superior: {limite_superior:.2f}")
print(f"Quantidade de Outliers encontrados: {len(outliers)}")
print("\nExemplos de registros considerados outliers:")
print(outliers.head())

Dataset 'Tips' carregado com sucesso!

--- Resultados do Exercício 3.1 ---
Q1: 13.35 | Q3: 24.13 | IQR: 10.78
Limite Superior: 40.30
Quantidade de Outliers encontrados: 9

Exemplos de registros considerados outliers:
     total_bill    tip     sex smoker   day    time  size
59        48.27   6.73    Male     No   Sat  Dinner     4
102       44.30   2.50  Female    Yes   Sat  Dinner     3
142       41.19   5.00    Male     No  Thur   Lunch     5
156       48.17   5.00    Male     No   Sun  Dinner     6
170       50.81  10.00    Male    Yes   Sat  Dinner     3


### Análise do Aluno:
[Encontrei 9 outliers na total_bill. Isso mostra que a maioria das contas fica abaixo de $40.30, mas temos alguns consumos bem acima da média. É importante isolar esses casos para não distorcer a análise geral do ticket médio.]

## 4. Assimetria (Skewness) e Curtose

A assimetria (skewness) mede o grau de inclinação de uma distribuição de probabilidade. Uma distribuição simétrica tem skewness zero. Valores positivos indicam uma cauda direita mais longa, enquanto valores negativos indicam uma cauda esquerda mais longa.

A curtose (kurtosis) mede o 'achatamento' ou 'pico' de uma distribuição. Uma curtose alta significa mais valores nas caudas e um pico mais nítido (leptocúrtica), enquanto uma curtose baixa indica menos valores nas caudas e um pico mais plano (platocúrtica). Uma distribuição normal tem curtose próxima de 3 (ou 0, se for calculada a curtose em excesso).

### Exercício 4.1: Cálculo e Interpretação de Assimetria e Curtose

Calcule a assimetria (skewness) e a curtose para as colunas numéricas ('total_bill', 'tip', 'size').

**Pergunta Dissertativa:** Com base nos valores calculados, descreva se a distribuição de cada variável numérica é simétrica, enviesada para a esquerda ou para a direita, e se ela é leptocúrtica, mesocúrtica ou platocúrtica. O que isso implica sobre a concentração dos dados?

In [6]:
import seaborn as sns
import pandas as pd

tips = sns.load_dataset('tips')

colunas_alvo = ['total_bill', 'tip', 'size']
dados_numericos = tips[colunas_alvo]

print(f"Assimetria: \n{dados_numericos.skew()}: ")
print(f"\nCurtose: \n{dados_numericos.kurtosis()}")

Assimetria: 
total_bill    1.133213
tip           1.465451
size          1.447882
dtype: float64: 

Curtose: 
total_bill    1.218484
tip           3.648376
size          1.731700
dtype: float64


### Análise do Aluno:
Bem, de acordo com os dados, as váriaveis `total_bill`, `tip` e `size` acabaram apresentando o seguinte comportamento:

**Assimetria (Skewness):**
- Aqui, todas possuem assimetria positiva, tendo os valores maiores que 0 (zero), indicando que suas distribuições são enviadas para a direita.

**Curtose (Kurtosis):**
- Aqui, todas possuem curtose em excesso positiva. sendo este classificado coomo leptocúrticas.

## 5. Correlação

A correlação mede a força e a direção da relação linear entre duas variáveis numéricas. Um coeficiente de correlação próximo de 1 indica uma forte relação linear positiva, -1 indica uma forte relação linear negativa, e 0 indica nenhuma relação linear. A matriz de correlação nos permite visualizar essas relações para todos os pares de variáveis numéricas em nosso dataset.

### Exercício 5.1: Matriz de Correlação

Calcule a matriz de correlação apenas para as features numéricas do dataset (`'total_bill'`, `'tip'`, `'size'`). Em seguida, visualize essa matriz utilizando um mapa de calor (heatmap) do seaborn.

In [ ]:
# DIGITE SEU CÓDIGO AQUI



### Análise do Aluno:
[Quais pares de variáveis possuem a correlação mais forte? Essa correlação é positiva ou negativa? O que esses resultados significam no contexto do dataset 'Tips'?]

## 6. Visualização

A visualização de dados é uma parte essencial da EDA. Gráficos como histogramas e boxplots nos permitem entender rapidamente a distribuição, dispersão e a presença de outliers, muitas vezes de forma mais intuitiva do que apenas com números.

### Exercício 6.1: Histogramas com KDE

Crie histogramas com a Estimativa de Densidade Kernel (KDE) para as colunas numéricas: 'total_bill', 'tip' e 'size'. Os histogramas mostram a frequência de valores em intervalos (bins), e o KDE suaviza essa distribuição, mostrando a densidade de probabilidade.

In [ ]:
# DIGITE SEU CÓDIGO AQUI



### Análise do Aluno:
[Observe a forma de cada distribuição. Elas parecem normais? Identifique se há assimetria visualmente. O que o KDE adiciona à sua compreensão da distribuição?]

### Exercício 6.2: Boxplots

Crie boxplots para as colunas numéricas: 'total_bill', 'tip' e 'size'. Os boxplots são excelentes para visualizar a mediana, quartis, dispersão e, crucialmente, identificar outliers visualmente.

In [ ]:
# DIGITE SEU CÓDIGO AQUI



### Análise do Aluno:
[Para cada boxplot, identifique a mediana (linha central), os quartis (caixa), o alcance dos dados (bigodes) e os outliers (pontos isolados). Compare esses outliers visuais com os que você calculou no Exercício 3.1 para 'total_bill'.]

## 7. Normalização de Dados

A normalização é um processo crucial no pré-processamento de dados, especialmente quando algoritmos de Machine Learning são sensíveis à escala das features (por exemplo, modelos baseados em distância como KMeans, SVMs com kernels, ou Redes Neurais). O objetivo é ajustar os valores das features para uma escala comum, sem distorcer as diferenças nas faixas de valores ou perder informações. Existem duas técnicas comuns:

1.  **Min-Max Scaling (Normalização)**: Escala os dados para um intervalo fixo, geralmente entre 0 e 1. A fórmula é: `X_novo = (X - X_min) / (X_max - X_min)`.
2.  **StandardScaler (Padronização - Z-score)**: Transforma os dados para ter média 0 e desvio padrão 1. A fórmula é: `X_novo = (X - média) / desvio_padrão`.

### Exercício 7.1: Aplicação de Normalização Min-Max e StandardScaler

Utilize as colunas numéricas ('total_bill', 'tip', 'size') para aplicar dois métodos de normalização:

1.  **Min-Max Scaling**: Normalize estas colunas para o intervalo [0, 1]. Armazene os resultados em um novo DataFrame ou adicione colunas normalizadas ao DataFrame `df` (ex: `total_bill_minmax`).
2.  **StandardScaler (Z-score)**: Padronize as mesmas colunas para ter média 0 e desvio padrão 1. Armazene os resultados em um novo DataFrame ou adicione colunas padronizadas ao DataFrame `df` (ex: `total_bill_zscore`).

Lembre-se de importar `MinMaxScaler` e `StandardScaler` do `sklearn.preprocessing`.

In [ ]:
# DIGITE SEU CÓDIGO AQUI

from sklearn.preprocessing import MinMaxScaler, StandardScaler

# Selecionando as colunas numéricas para normalização


# 1. Min-Max Scaling



# 2. StandardScaler (Z-score)



### Análise do Aluno:
[Compare os valores originais com os valores normalizados por Min-Max e padronizados por StandardScaler. Observe as diferenças na escala e no range dos dados. Quando seria mais apropriado usar um método em vez do outro?]